src/Commands/core/ArchiveRestoreCommands.php

```diff
@@ -0,0 +1,489 @@
+            $this->importFiles($filesComponentPath, $site);
+        }
+
+        if ($options['db']) {
+            $databaseComponentPath = $options['db_path'] ?? Path::join($extractDir, self::COMPONENT_DATABASE, self::SQL_DUMP_FILE_NAME);
+            $this->importDatabase($databaseComponentPath, $options);
+        }
+
+        $this->logger()->info(dt('Done!'));
+    }
+
+    /**
+     * Creates a temporary directory to extract the archive onto.
+     *
+     * @throws \Exception
+     */
+    protected function prepareTempDir(): void
+    {
+        $this->filesystem = new Filesystem();
+        $this->extractedPath = FsUtils::prepareBackupDir(self::TEMP_DIR_NAME);
+        register_shutdown_function([$this, 'cleanUp']);
+    }
+
+    /**
+     * Extracts the archive.
+     *
+     * @param string $path
+     *   The path to the archive file.
+     * @param array $options
+     *   Command options.
+     *
+     * @return string
+     *
+     * @throws \Exception
+     */
+    protected function extractArchive(string $path, array $options): string
+    {
+        $this->logger()->info('Extracting the archive...');
+
+        if (!is_file($path)) {
+            throw new Exception(dt('File !path is not found.', ['!path' => $path]));
+        }
+
+        if (!preg_match('/\.tar\.gz$/', $path)) {
+            throw new Exception(dt('File !path is not a *.tar.gz file.', ['!path' => $path]));
+        }
+
+        ['filename' => $archiveFileName] = pathinfo($path);
+        $archiveFileName = str_replace('.tar', '', $archiveFileName);
+
+        $extractDir = Path::join(dirname($path), $archiveFileName);
+        if (is_dir($extractDir)) {
+            if ($options['overwrite']) {
+                $this->filesystem->remove($extractDir);
+            } else {
+                throw new Exception(
+                    dt('Extract directory !path already exists (use "--overwrite" option).', ['!path' => $extractDir])
+                );
+            }
+        }
+
+        $this->filesystem->mkdir($extractDir);
+
+        $archive = new PharData($path);
+        $archive->extractTo($extractDir);
+
+        $this->logger()->info(dt('The archive successfully extracted into !path', ['!path' => $extractDir]));
+
+        return $extractDir;
+    }
+
+    /**
+     * Imports the code to the site.
+     *
+     * @param string $source
+     *   The path to the code files directory.
+     * @param string|null $destinationSite
+     *   The destination site alias.
+     *
+     * @throws \Drush\Exceptions\UserAbortException
+     * @throws \Exception
+     */
+    protected function importCode(string $source, ?string $destinationSite): void
+    {
+        $this->logger()->info('Importing code...');
+
+        if (!is_dir($source)) {
+            throw new Exception(dt('Directory !path not found.', ['!path' => $source]));
+        }
+
+        $siteAlias = $this->getSiteAlias($destinationSite);
```

Comment:
If the cwd is not within a Drupal site (e.g. cwd is inside an empty directory), then we should create a new site and extract the files there.